nextcord/http.py

```diff
@@ -155,25 +164,280 @@ def bucket(self) -> str:
+    """Used to time gate a large batch of requests to only occur X every Y seconds. Used via ``async with``
+
+    NOT THREAD SAFE.
+
+    Parameters
+    ----------
+    time_offset: :class:`float`
+        Number in seconds to increase all timers by. Used for lag compensation.
+    """
+
+    def __init__(self, time_offset: float = 0.3) -> None:
+        self.limit: int = 1
+        """Maximum amount of requests before requests have to wait for the rate limit to reset."""
+        self.remaining: int = 1
+        """Remaining amount of requests before requests have to wait for the rate limit to reset."""
+        self.reset: datetime | None = None
+        """Datetime that the bucket roughly will be reset at."""
+        self.reset_after: float = 1.0
+        """Amount of seconds roughly until the rate limit will be reset."""
+        self.bucket: str | None = None
+        """Name of the bucket, if it has one."""
+
+        self._time_offset: float = time_offset
+        """Number in seconds to increase all timers by. Used for lag compensation."""
+        self._first_update: bool = True
+        """If the next update to be ran will be the first."""
+        self._reset_remaining_task: asyncio.Task | None = None
+        """Holds the task object for resetting the remaining count."""
+        self._on_reset_event: asyncio.Event = asyncio.Event()
+        """Used to indicate when the rate limit is ready to be acquired."""
+        self._on_reset_event.set()
+        self._deny: bool = False
+        """Set to error all acquiring requests with a 404 value error."""
+        self._migrating: str | None = None
+        """When this RateLimit is being deprecated and acquiring requests need to migrate to a different RateLimit, this
+        variable should be set to the different RateLimit/buckets string name.
+        """
+
+    @property
+    def resetting(self) -> bool:
+        return self._reset_remaining_task is not None and not self._reset_remaining_task.done()
+
+    async def update(self, response: aiohttp.ClientResponse) -> None:
+        """Updates the rate limit with information from the response."""
+
+        if response.headers.get("X-RateLimit-Global") == "true":
+            # The response is intended for the global rate limit, not a regular rate limit.
+            return
+
+        # Updates the bucket name. The bucket name not existing as fine, as ``None`` is desired for that.
+        # This is done immediately, so we can error out if we get an update not for this bucket.
+        x_bucket = response.headers.get("X-RateLimit-Bucket")
+
+        if self.bucket == x_bucket:
+            pass  # Don't need to set it again.
+        elif self.bucket is None:
+            self.bucket = x_bucket
+        else:
+            raise IncorrectBucket(
+                f"Update given for bucket {x_bucket}, but this RateLimit is for bucket {self.bucket}!"
+            )
+
+        if response.status == 404:
+            self._deny = True
```

Comment:
I don't like this automatic denial system. If a bot repeatedly calls the same request with the same parameters that already 404ed, then that's the fault of the bot for not properly handling those 404s. Nextcord shouldn't be responsible for that.

Reply:
IIRC, this was added with `asyncio.gather()` in mind. In the event that a large batch of requests are gathered up and a mispelling in the route was made, it would cancel almost all of them to avoid Discord getting upset and Cloudflare possibly taking out the ban hammer. It seemed to me that it was a relatively free check that could save a major headache for some people. I can remove if though if you wish.

Reply:
Updated it so a better, less generic error is raised.

Reply:
That's not the problem I was addressing. I think this feature should be toggleable, and it should be off by default. Not everyone is going to be using `asyncio.gather()`, after all.

Reply:
After a small talk on the Discord, it was decided that removing the denial system was a better idea than adding a toggle for it, with it toggled off by default.